service/googlechat/googlechat.go

```diff
@@ -0,0 +1,97 @@
+// Package googlechat provides message notification integration sent to multiple
+// spaces within a Google Chat Application.
+package googlechat
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"google.golang.org/api/chat/v1"
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/option"
+)
+
+//go:generate mockery --name=spacesMessageCreator --output=. --case=underscore --inpackage
+type spacesMessageCreator interface {
+	Create(string, *chat.Message) createCall
+}
+
+//go:generate mockery --name=createCall --output=. --case=underscore --inpackage
+type createCall interface {
+	Do(...googleapi.CallOption) (*chat.Message, error)
+}
+
+// Compile-time check to ensure that client implements the spaces message service
+// interface.
+var (
+	_ spacesMessageCreator = new(messageCreator)
+	// Compile-time check to ensure that client implements the create call interface.
+	_ createCall = new(chat.SpacesMessagesCreateCall)
+)
+
+// messageCreator is wrapper struct for the native chat.SpacesMessagesService struct.
+// This exists so that we can mock the chat.SpacesMessagesCreateCall with the common
+// interface "createCall".
+type messageCreator struct {
+	*chat.SpacesMessagesService
+}
+
+func newMessageCreator(options ...option.ClientOption) (spacesMessageCreator, error) {
+	ctx := context.Background()
+	svc, err := chat.NewService(ctx, options...)
```

Comment:
The context should be controlable from a higher level. Could you adapt this, so that the context gets passed to this function from higher-level functions?

Thinking about how to solve this right now. My issue is the constructor function `New()`, which utilizes this function to setup the message creator. So passing a context to `New()` would be easy enough, however, this breaks consistency with all other service constructor functions..

That made me think, do we even need to create the `spacesMessageCreator` in the services constructor function, or, would it be sufficient to create it on premise in the `Send()` method? In `Send()` we naturally have an existing `context.Context` already available too.

What do you think, @EthanEFung? You know the ins and outs of this service much better than I do.

Reply:
I can certainly explore making the changes to the service. Here are my thoughts about how I'd go about it.

1. Replace the `messageCreator` attribute  with a `clientOptions` attribute. This would be necessary so that whatever authentication strategy the notify user wanted to use, it would still be available.

```go
type Service struct {
    options []option.ClientOption
    spaces  []string
}
```
2. Change the function signature for `newMessageCreator` to accept a context
2. Instead of creating the messageCreator when instantiating the service, we create a new messageCreator every time the `Send` method is invoked as suggested:
```go
func (s Service) Send(ctx context.Context, subject, message string) error {
  svc, err := newMessageCreator(ctx, ...m.options)
  if err != nil {
    return err
  }
  // the rest of the send logic ...
}
```
The only one drawback I can think of right now is that can't know whether the authentication strategy is valid when the service is instantiated, and have to wait until a message is sent.

Reply:
@EthanEFung thanks for that, makes sense. I went through their code and there's no way avoiding this. They're initialzing the `http.Client` with the context and at no point expose that client downstream.

Our options I see right now:

1. We pick the solution discussed above.
2. We're keeping your initial solution but expand the package with a `NewWithContext(context.Context, ...)` constructor.

I'm leaning to solution no. 2. Constructor functions like this are a relatively common pattern in Notify already, I think, we'd be fine doing this. Let me know what you think.

Reply:
I'm fine with option 2. Looking at the code I think it would be relatively easy to implement this way.